src/lib/optimizer/join_graph.hpp

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "optimizer/abstract_syntax_tree/abstract_ast_node.hpp"
+#include "optimizer/abstract_syntax_tree/predicate_node.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+class JoinNode;
+class PredcicateNode;
+
+/**
+ * A connection between two JoinGraph-Vertices.
+ */
+struct JoinEdge {
+  JoinEdge(const std::pair<JoinVertexID, JoinVertexID>& vertex_indices, JoinMode join_mode);
+  JoinEdge(const std::pair<JoinVertexID, JoinVertexID>& vertex_indices, const std::pair<ColumnID, ColumnID>& column_ids,
+           JoinMode join_mode, ScanType scan_type);
+
+  std::pair<JoinVertexID, JoinVertexID> vertex_indices;
+  std::optional<std::pair<ColumnID, ColumnID>> column_ids;
+  JoinMode join_mode;
+  std::optional<ScanType> scan_type;
+};
+
+struct JoinVertexPredicate {
+  JoinVertexPredicate(ColumnID column_id, ScanType scan_type, const AllParameterVariant& value);
+
+  ColumnID column_id;
+  ScanType scan_type;
+  AllParameterVariant value;
+};
+
+struct JoinVertex {
+  explicit JoinVertex(const std::shared_ptr<AbstractASTNode>& node);
+
+  std::shared_ptr<AbstractASTNode> node;
+  std::vector<JoinVertexPredicate> predicates;
```

Comment:
When do we use the word `predicate` and when do we use `edge`?

Does a `JoinVertex` connect always two `JoinNodes`? What is the maximum number of `predicates` then?